src/components/swap/ConfirmSwapModal.tsx

```diff
@@ -80,48 +81,85 @@ function useConfirmModalState({
       steps.push(ConfirmModalState.PERMITTING)
     }
     steps.push(ConfirmModalState.PENDING_CONFIRMATION)
-    setPendingModalSteps(steps)
+    return steps
   }, [allowance])
 
   const { chainId } = useWeb3React()
   const trace = useTrace()
   const maximumAmountIn = useMaxAmountIn(trade, allowedSlippage)
 
-  const startSwapFlow = useCallback(async () => {
-    setApprovalError(undefined)
-    if (allowance.state === AllowanceState.REQUIRED) {
-      // Starts the approval process, by triggering either the Token Approval or the Permit signature.
-      try {
-        if (allowance.needsSetupApproval) {
+  const catchUserReject = async (action: () => Promise<void>, onRealError: () => void) => {
+    try {
+      await action()
+    } catch (e) {
+      setConfirmModalState(ConfirmModalState.REVIEWING)
+      if (didUserReject(e)) {
+        return
+      }
+      console.error(e)
+      onRealError()
+    }
+  }
+
+  const performStep = useCallback(
+    async (step: ConfirmModalState) => {
+      switch (step) {
+        case ConfirmModalState.APPROVING_TOKEN:
           setConfirmModalState(ConfirmModalState.APPROVING_TOKEN)
-          await allowance.approve()
-          sendAnalyticsEvent(InterfaceEventName.APPROVE_TOKEN_TXN_SUBMITTED, {
-            chain_id: chainId,
-            token_symbol: maximumAmountIn?.currency.symbol,
-            token_address: maximumAmountIn?.currency.address,
-            ...trace,
-          })
-        } else {
+          invariant(allowance.state === AllowanceState.REQUIRED, 'Allowance should be required')
+          await catchUserReject(
+            async () => {
+              await allowance.approve()
+              sendAnalyticsEvent(InterfaceEventName.APPROVE_TOKEN_TXN_SUBMITTED, {
+                chain_id: chainId,
+                token_symbol: maximumAmountIn?.currency.symbol,
+                token_address: maximumAmountIn?.currency.address,
+                ...trace,
+              })
+            },
+            function onRealError() {
+              setApprovalError(PendingModalError.TOKEN_APPROVAL_ERROR)
+            }
+          )
+          break
```

Comment:
I think this would potentially work better w/ the updated implementation of catchUserReject above:
```suggestion
          allowance.approve()
            .then(sendAnalytics)
            .catch((e) => catchUserReject(e, PendingModalError.TOKEN_APPROVAL_ERROR))
          break
```

Reply:
(and do similar for the steps)

Reply:
i like it